v2/internal/testcommon/replay_roundtripper.go

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package testcommon
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"gopkg.in/dnaeon/go-vcr.v3/cassette"
+)
+
+// replayRoundTripper wraps an inner round tripper and replays requests in order to improve the resilience of ASO tests.
+// PUT requests are cached by target URL, and may be replayed ONCE if an extra PUT occurs.
+// GET requests are cached by target URL, and may be replayed multiple times.
+// This combination should allow additional reconciles - an extra PUT gets returned the same long running operation as
+// the original, which a GET then shows is complete.
+type replayRoundTripper struct {
+	inner   http.RoundTripper
+	gets    map[string]*http.Response
+	puts    map[string]*http.Response
+	log     logr.Logger
+	padlock sync.Mutex
+}
+
+var _ http.RoundTripper = &replayRoundTripper{}
+
+// newReplayRoundTripper creates a new replayRoundTripper that will replay selected requests to improve test resilience.
+func newReplayRoundTripper(
+	inner http.RoundTripper,
+	log logr.Logger,
+) *replayRoundTripper {
+	return &replayRoundTripper{
+		inner: inner,
+		gets:  make(map[string]*http.Response),
+		puts:  make(map[string]*http.Response),
+		log:   log,
+	}
+}
+
+// RoundTrip implements http.RoundTripper.
+func (replayer *replayRoundTripper) RoundTrip(request *http.Request) (*http.Response, error) {
+	if request.Method == http.MethodGet {
+		return replayer.roundTripGet(request)
+	}
+
+	if request.Method == http.MethodPut {
+		return replayer.roundTripPut(request)
+	}
+
+	// For other kinds of request, just pass through to the inner round tripper.
+	return replayer.inner.RoundTrip(request)
+}
+
+func (replayer *replayRoundTripper) roundTripGet(request *http.Request) (*http.Response, error) {
+	// First use our inner round tripper to get the response.
+	response, err := replayer.inner.RoundTrip(request)
+	if err != nil {
+		// We have an error - return it, unless it's from go-vcr
+		if !errors.Is(err, cassette.ErrInteractionNotFound) {
```

Comment:
Count's are independent for GET vs PUT for the same URL (see the initialization of the variable `key` on line 38 of  `counting_roundtripper.go`).

That said, the `COUNT` header is irrelevant because this only triggers once we fall off the end of the recording - it doesn't matter that a subsequent GET specifies COUNT+1 because there are no more recorded responses for that combination of method and URL.

The sequence of requests from a single reconciler is stable, which is why we didn't have problems doing replay of simple resources, only when we got to more complex scenarios where multiple reconcilers were running in parallel.

Different reconcilers are handling different resources, which have different URLs, so I don't think the scenario you described above can happen.

We'd actually have the following:

1. PUT (COUNT=0) for create, served from the recording file
2. GET (COUNT=0) for monitoring the long running operation
3. GET (COUNT=1...n) for monitoring the LRO until it completes
4. PUT (COUNT=1) for update, served from the recording file
5. GET (COUNT=0) for monitoring a new long running operation
6. GET (COUNT=1...n) for monitoring the LRO until it completes
7. PUT (COUNT=2) for extra update, no available recording, replay from COUNT=1, returning the same LRO
8. GET (COUNT=n+1) for monitoring the LRO, no available recording, replay from COUNT=n, indicating the LRO has completed

I can add the specHash if that's needed (I have it in another working branch) but I'm not sure it is.

Have I covered off your concerns, or is there something I need to address?

Reply:
> Different reconcilers are handling different resources, which have different URLs, so I don't think the scenario you described above can happen.

The scenario I described above is for the same resource, same name, same type, where the first PUT is a create, and the second PUT is an update.

Imagine I have test code like this:
```
// create my StorageAccount
acct := &storage.StorageAccount{
	ObjectMeta: tc.MakeObjectMetaWithName(tc.NoSpaceNamer.GenerateName("stor")),
	Spec: storage.StorageAccount_Spec{
		Location: tc.AzureRegion,
		Owner:    testcommon.AsOwner(rg),
		Kind:     to.Ptr(storage.StorageAccount_Kind_Spec_StorageV2),
		Sku: &storage.Sku{
			Name: to.Ptr(storage.SkuName_Standard_LRS),
		},
		// TODO: They mark this property as optional but actually it is required
		AccessTier: to.Ptr(storage.StorageAccountPropertiesCreateParameters_AccessTier_Hot),
	},
}

// Triggers duplicate reconcile during recordings 5% of the time and consumes 2 PUTs.
// The second PUT is for the update request below, and the LRO returned is for that as well, 
// so at the end of the LRO the acct variable 5% of the time will have the second LROs response (after update)
// and 95% of the time will have the first LROs response (before update).
tc.CreateResourceAndWait(acct)

// This assert will fail 5% of the time
tc.Expect(acct.Status.AccessTier).To(Equal(to.Ptr(storage.StorageAccountPropertiesCreateParameters_STATUS_AccessTier_Hot)))

old := acct.DeepCopy()
acct.Spec.AccessTier = to.Ptr(storage.StorageAccountPropertiesCreateParameters_AccessTier_Cold),
tc.PatchResourceAndWait(old, &acct)

tc.Expect(acct.Status.AccessTier).To(Equal(to.Ptr(storage.StorageAccountPropertiesCreateParameters_STATUS_AccessTier_Hot)))
```

You're right that I forgot about LROs but I don't think it saves us from the above?

Now, I don't know how often that the tests we have today actually do this pattern of:
```
// create
// assert propA has shape 1
// update
// assert propA has shape 2
```

but I do think there are a few, and it would be easy to write more that (mostly) passed but flaked occasionally.